### ISSUE TITLE
dpkt.Writer does not write packet trailer

### ISSUE THREAD

**Issue by mikewalshchicago** [kbandla/dpkt#438]
I'm new to GitHub so I apologize if this is not the correct place to ask questions, but i could not find anything regarding packet trailers in the documentation.

I am using pcapy.open_live() to read the packets from a bytestream.  I am then using dpkt to do some read-only analysis, and then writing each packet to a pcap file using dpkt.Writer.  An abstract of what my code is doing is as follows:

#first, i'm opening the bytestream, passing the desired interface from cli arguments.  I have tried larger #snaplen values however I saw no difference 
cap = pcapy.open_live(interface, 65536, 1, 0)

#i want to write every UDP message in this packet capture, so i do the following:
while True:
(header, packet) = cap.next()
ethernet = dpkt.ethernet.Ethernet(packet)

   if ethernet.type == dpkt.ethernet.ETH_TYPE_IP:
      ip = ethernet.data
      if ip.p == dpkt.ip.IP_PROTO_UDP:
          udp = ip.data
#there's some processing that goes on in the middle, but i am not manipulating the output. I have tried #different snaplen arguments here as well, e.g. snaplen=65536, but see no difference in the output
   writer = dpkt.pcap.Writer(open(path + filename)+'.pcap', 'wb+'),nano=True)
#I have also tried to write 'ip' as ouput, which appears the same.  
   writer.writepkt(ethernet)

The headers and payloads look fine in my pcap, but each message also has an 16 byte packet trailer which i do not see. if I use tcpdump -w, no special arguments are needed to see the trailer. 

I'd appreciate any help you can give me, and happy to help where i can with diagnosis

**Comment by obormot** [comment]
Please attach a sample pcap - even 1 packet with the 16-byte trailer would help to reproduce the issue. I think dpkt may not be preserving the padding / trailer in some cases.

**Comment by mikewalshchicago** [comment]
Please see attached, and thanks for the quick reply!

[trailer.zip](https://github.com/kbandla/dpkt/files/2705353/trailer.zip)

**Comment by obormot** [comment]
Confirmed. Interestingly enough there's a single IP packet in the pcap and exactly this packet gets packaged w/o the 16 bytes trailer!
```
In [31]: str(eth) == pp[1]
Out[31]: False

In [32]: len( str(eth) )
Out[32]: 243

In [33]: len(pp[1])
Out[33]: 259

# difference is 16 bytes
```

On another note, dpkt decodes the trailer / padding and FCS incorrectly for these packets as well. Second packet:
```
In [20]: hexdump(p0)
0000   01 80 C2 00 00 00 78 0C  F0 B4 D8 91 00 27 42 42   ......x......'BB
0010   03 00 00 02 02 3C 00 01  2C 33 11 F2 39 C1 00 00   .....<..,3..9...
0020   00 02 80 01 78 0C F0 B4  D8 BC 80 AA 01 00 14 00   ....x...........
0030   02 00 0F 00 00 00 00 00  00 00 00 00 4D B9 81 20   ............M..
0040   5C 1E 5F BA 3A A5 47 FA  01 8E 52 03               \._.:.G...R.

In [21]: Ethernet(p0)
Out[21]: Ethernet(
  dst='\x01\x80\xc2\x00\x00\x00',   
  src='x\x0c\xf0\xb4\xd8\x91', 
  type=39, 
  fcs=0x018e5203,      
    # incorrect; picked last 4 bytes for FCS instead of 4 bytes following the padding 
  trailer='\x00\x00\x00\x00\x00\x00\x00M\xb9\x81 \\\x1e_\xba:\xa5G\xfa',  
    # incorrect; first 7 zeroes should be padding;  followed by FCS, followed by eth.trailer (Metamako)
  len=39, 
  data=LLC(dsap=66, ssap=66, data=STP(v=2, type=2, flags=60, root_id='\x00\x01,3\x11\xf29\xc1', root_path=2, bridge_id='\x80\x01x\x0c\xf0\xb4\xd8\xbc', port_id=32938, age=1, age=1, hello=2, fd=15, data='\x00')))
```

**Comment by mikewalshchicago** [comment]
Oscar,

Thanks for the proposed fix.  I cloned this repo and made the changes from your commit in the ip.py file, and then compiled it with "python setup.py install"  i uninstalled the dpkt i was previously using with pip, so the fact that my code ran at all makes me think i did the proper procedure to test.  I'm sorry to report that i still do not see the trailer on the messages in wireshark, where i do in a corresponding tcpdump.  

Please let me know if there is anything else I can try to help get us closer to a fix

**Comment by mikewalshchicago** [comment]
oh, i should not leave out that I need to add some exception handling after removing the logic from ip.py.  the stacktrace itself may be a clue to you:  

NeedData
    dpkt.Packet.__init__(self, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/dpkt-1.9.2-py2.7.egg/dpkt/dpkt.py", line 90, in __init__
    self.unpack(args[0])
  File "/usr/lib/python2.7/site-packages/dpkt-1.9.2-py2.7.egg/dpkt/ethernet.py", line 168, in unpack
    self.data = self.llc = llc.LLC(self.data[:eth_len])
  File "/usr/lib/python2.7/site-packages/dpkt-1.9.2-py2.7.egg/dpkt/dpkt.py", line 93, in __init__
    raise NeedData

**Comment by mikewalshchicago** [comment]
Any updates or suggestions on this?  I'm still eager to assist with a resolution

**Comment by mikewalshchicago** [comment]
Oscar,

For my purposes every message in the pcap will  have the 16 byte trailer.  Theoretically, if I added 16 bytes to the self.len value, would this work?  

I don't see why it wouldn't unless dpkt will reject/discard a binary string that doesn't align with the checksum of the packet.

**Comment by obormot** [comment]
I may have some time next week to look into it. The patch I suggested fixes the issue in the pcap you attached. If there's still an issue please provide steps to reproduce. On a side note, manipulating the IP len in dpkt has been a grey area for a long time; I don't think it's working great at the moment.

**Comment by mikewalshchicago** [comment]
Oscar,

Thanks for the reply.  It's possible I did not correctly compile your
changes, however I did as described in the comments for the raised issue:

1. Cloned the repo
2. Edited the file as I saw in the tagged fix (removed the depreciated
decorator and I believe there was one other line)
3. Uninstalled don't from pip
4. Compiled with setup.py install
5. Ran my code as well as a tcpdump. Trailer was missing from my written
file but present in pcap.

Let me try the steps above again, but provide more specifics if you don't
see anything wrong with the sequence of events above.

I will also try my hand at changing IP len before next week.  I can
definitely see why doing so would cause issues if my offset overruns the
byte string.

**Comment by mikewalshchicago** [comment]
So, i apologize for the formatting, but i'm in a bit of a rush.  I attempted to compile dpkt with your commit but had the result i did the first time i tried.  here are my steps in more detail that i provided before:

environment - python 2.7
OS - centos 7.6

steps taken:
1. git clone https://github.com/kbandla/dpkt.git
2. vi dpkt/dpkt/ip.py
3. delete following lines:
    
	from .decorators import deprecated
	
	        if self.len:
            buf = buf[self.__hdr_len__ + ol:self.len]
        else:  # very likely due to TCP segmentation offload
            buf = buf[self.__hdr_len__ + ol:]
			
4. change indentation on following line:

   buf = buf[self.__hdr_len__ + ol:]
   
 5. uninstall dpkt pip package
 DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
Uninstalling dpkt-1.9.2:
  Would remove:
    /usr/lib/python2.7/site-packages/dpkt-1.9.2-py2.7.egg-info
    /usr/lib/python2.7/site-packages/dpkt/*
6. compile modified source with 'sudo python setup.py install'. no errors on compilation.  here is the trailing output:

creating dist
creating 'dist/dpkt-1.9.2-py2.7.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing dpkt-1.9.2-py2.7.egg
creating /usr/lib/python2.7/site-packages/dpkt-1.9.2-py2.7.egg
Extracting dpkt-1.9.2-py2.7.egg to /usr/lib/python2.7/site-packages
Adding dpkt 1.9.2 to easy-install.pth file

Installed /usr/lib/python2.7/site-packages/dpkt-1.9.2-py2.7.egg
Processing dependencies for dpkt==1.9.2
Finished processing dependencies for dpkt==1.9.2

7. ran a stripped down version of my code.  Essentially, a single function that uses a command line argument for interface name


####MW Writer code independent of generator

def pcapwriter(interface):
    print ("debug: pcap writer started")
#MW create writer dict and file rotate counter
    dpktwrite=dpkt.pcap.Writer(open('/tmp/dpkt.pcap', 'wb+'))
    cap = pcapy.open_live(interface, 65536, 1, 1)
    while True:
       (header, packet) = cap.next()
       ethernet = dpkt.ethernet.Ethernet(packet)
       dpktwrite.writepkt(ethernet)
	   
8. dpkt.pcap file is readable by wireshark version 2.6.3 (windows 64 bit), but there are no trailers on any packet  dpkt.pcap attached

[dpkt.zip](https://github.com/kbandla/dpkt/files/2867513/dpkt.zip)
[tcpdump.zip](https://github.com/kbandla/dpkt/files/2867514/tcpdump.zip)

9. i ran a tcpdump on the same interface, capturing only tcp and writing a file called /tmp/tcpdump.pcap.  surprisingly the file from the same time is orders of magnitude larger than the dpkt one, even though tcpdump looks to be mostly ethernet frames.

sudo ./dpkttest.py -i enp94s0f0 & sudo tcpdump -i enp94s0f0 -w /tmp/tcpdump.pcap &

tcpdump version 4.9.2
libpcap version 1.8.1

(i also created a smaller file to upload after the fact so you don't have to download a 44mb file to see that there are packet trailers. 

sudo tcpdump -i enp94s0f0 tcp -w /tmp/quick.pcap

[quick.zip](https://github.com/kbandla/dpkt/files/2867517/quick.zip)

so TL;DR: I am either missing your proposed changes, or the changes dont seem to have any effect on my particular issue.  Hope this is more informative than my previous posts.  I'm standing by if there's anything else i can do

**Comment by obormot** [comment]
Ok, this is useful to rule out if dpkt is to blame, or pcapy being part of the issue.  I suggest trying the following experiment: skip dpkt's Ethernet() conversion and write pcapy's buffers directly. I.e. modifying your code fragment:
```
while True:
    (header, packet) = cap.next()
    # disabled    ethernet = dpkt.ethernet.Ethernet(packet)
    dpktwrite.writepkt(packet)   # write packet directly
```
In this scenario if you open the resulting pcap file in Wireshark, does it have the trailers?

**Comment by mikewalshchicago** [comment]
Oscar,  am writing packet instead of ethernet as you suggested and the trailer is still not seen in wireshark.  I take it that this is suggests that the issue is with pcapy.  do you have any other suggestions for what i could replace pcapy with to open the live bytestream and confirm that pcapy is indeed stripping the trailer?

**Comment by obormot** [comment]
In the past I've used nfqueue python bindings to sniff and pass/block traffic. 
Not sure which package I used bad then, but a quick search reveals this example: https://pypi.org/project/NetfilterQueue/

**Comment by mikewalshchicago** [comment]
Oscar,

Great news:  I replaced the pcapy module with pypcap and I am now getting the timestamp trailer when i write (packet) with dpkt.writer.  I have not yet confirmed if this is in combination with the changes you proposed (meaning i'm still running with the source compiled dpkt and not the one from the pip repo), but i will do that soon enough.  Thanks for all of your help with this problem!

**Comment by brifordwylie** [comment]
@mikewalshchicago @obormot I believe this issue has been resolved with the merging of #470, so I'll close this issue. If the issue is still unresolved, please comment here and we'll reopen.
